src/docker-image-deployment.ts

```diff
@@ -0,0 +1,123 @@
+
+  constructor(scope: Construct, id: string, props: DockerImageDeploymentProps) {
+    super(scope, id);
+
+    const handlerRole = new iam.Role(this, 'DockerImageDeployRole', {
+      assumedBy: new iam.ServicePrincipal('codebuild.amazonaws.com'),
+    });
+
+    const sourceConfig = props.source.bind(this, { handlerRole });
+    props.destination.bind(handlerRole);
+
+    const sourceUri = sourceConfig.imageUri;
+
+    const destTag = props.destination.config.destinationTag ?? sourceConfig.imageTag;
+    const destUri = `${props.destination.config.destinationUri}:${destTag}`;
+
+    const accountId = Stack.of(this).account;
+    const region = Stack.of(this).region;
+
+    const sourceLoginCommands = [
+      `aws ecr get-login-password --region ${region} | docker login --username AWS --password-stdin ${accountId}.dkr.ecr.${region}.amazonaws.com`,
+    ];
+
+    const buildCommands = [
+      `docker pull ${sourceUri}`,
+      `docker tag ${sourceUri} ${destUri}`,
+      `docker push ${destUri}`,
+    ];
+
+    this.cb = new codebuild.Project(this, 'DockerImageDeployProject', {
+      buildSpec: codebuild.BuildSpec.fromObject({
+        version: '0.2',
+        phases: {
+          pre_build: {
+            commands: sourceLoginCommands,
+          },
+          build: {
+            commands: buildCommands,
+          },
+        },
+      }),
+      environment: {
+        privileged: true,
+        buildImage: codebuild.LinuxBuildImage.STANDARD_5_0,
+      },
+      role: handlerRole,
+    });
+
+    const onEventHandler = new lambda.NodejsFunction(this, 'onEventHandler', {
+      entry: 'lib/codebuild-handler/index.js',
+      handler: 'onEventhandler',
+      depsLockFilePath: 'yarn.lock',
+      runtime: Runtime.NODEJS_16_X,
+    });
+
+    const isCompleteHandler = new lambda.NodejsFunction(this, 'isCompleteHandler', {
+      entry: 'lib/codebuild-handler/index.js',
+      handler: 'isCompleteHandler',
+      depsLockFilePath: 'yarn.lock',
+      runtime: Runtime.NODEJS_16_X,
+    });
+
+    iam.Grant.addToPrincipal({
+      grantee: onEventHandler,
+      actions: ['codebuild:StartBuild'],
+      resourceArns: [this.cb.projectArn],
+      scope: this,
+    });
+
+    iam.Grant.addToPrincipal({
+      grantee: isCompleteHandler,
+      actions: [
+        'codebuild:ListBuildsForProject',
+        'codebuild:BatchGetBuilds',
+      ],
+      resourceArns: [this.cb.projectArn],
+      scope: this,
+    });
+
+    const crProvider = new cr.Provider(this, 'CRProvider', {
+      onEventHandler: onEventHandler,
+      isCompleteHandler: isCompleteHandler,
+      queryInterval: Duration.seconds(30),
+      totalTimeout: Duration.minutes(10),
+    });
+
+    const customResource = new CustomResource(this, `CustomResource${Date.now().toString()}`, {
+      serviceToken: crProvider.serviceToken,
+      properties: {
+        projectName: this.cb.projectName,
+      },
+    });
+
+    customResource.node.addDependency(handlerRole);
```

Comment:
The grants of the roles governing the CR or the role assumed by the codebuild? Or both? Im assuming the grants for the lambdas. The code for the codebuild is dependent on the bind method (at least for Source).

Now that im looking at this, I wonder if the caller needs to call bind on both the Source and Destination. Right now it only has to on Source in order to get the properties it needs (at least the error would occur after synth).

Reply:
The question here is what does the custom resource depend on. Right now you have it depending on `handlerRole` and implicitly on the codebuild job (due to the `projectName` which is an attribute returned by Cfn. I'm 90% sure about this)., The `handler role` itself is implicitly depended on by the codebuild job. 

```mermaid
flowchart LR
    A[custom resource]-->B[handler role]
    C[codebuild]-->B
    A-->C
    D[grants]-->C
```

Note that the grants depend on codebuild since they use the codebuild arn. And that depending on the handler role is useless, since we already depend on codebuild.

The question is, do you need to depend on the grants. The answer is yes. Without them, your custom resource doesn't have the correct permissions to do what you need it to do. You're not currently doing that, so you're getting lucky. It's possible that sometimes, the custom resource executes before the grants, causing your lambda to fail. The dependency tree you want looks like this:

```mermaid
flowchart LR
    A[custom resource]-->D[grants]
    C[codebuild]-->B[handler role]
    D[grants]-->C
    A-->C
```

I added custom resource to codebuild still because that is an implicit dependency. Note that, by depending on the grants, you already are depending on the handler role.